### PR TITLE
Fix galore lr display with schedulers

### LINKED ISSUE #31707: When I used galore, the learning rate was set to 8e-6, but the training rate was 0.001
```
import os
import sys
from typing import List

import fire
import torch
import transformers
from datasets import load_dataset
import os

# os.environ["NCCL_P2P_DISABLE"] = "1"
# os.environ["NCCL_IB_DISABLE"] = "1"

"""
Unused imports:
import torch.nn as nn
import bitsandbytes as bnb
"""


from transformers import AutoModelForCausalLM, AutoTokenizer
from transformers import BitsAndBytesConfig
from peft import (
    LoraConfig,
    get_peft_model,
    prepare_model_for_kbit_training,
    set_peft_model_state_dict,
)
#from utils.prompter import Prompter
import signal
import sys
import os
os.environ["WANDB_DISABLED"] = "true"
def train(
    # model/data params
    base_model: str = "",  # the only required argument
    data_path: str = "yahma/alpaca-cleaned",
    output_dir: str = "./lora-alpaca",
    # training hyperparams
    batch_size: int = 128,
    micro_batch_size: int = 4,
    num_epochs: int = 3,
    learning_rate: float = 3e-4,
    cutoff_len: int = 256,
    val_set_size: int = 2000,
    # llm hyperparams
    train_on_inputs: bool = True,  # if False, masks out inputs in loss
    add_eos_token: bool = False,
    group_by_length: bool = False,  # faster, but produces an odd training loss curve
    # wandb params
    wandb_project: str = "",
    wandb_run_name: str = "",
    wandb_watch: str = "",  # options: false | gradients | all
    wandb_log_model: str = "",  # options: false | true
    resume_from_checkpoint: str = None,  # either training checkpoint or final adapter
    prompt_template_name: str = "alpaca2",  # The prompt template to use, will default to alpaca.
):
    if int(os.environ.get("LOCAL_RANK", 0)) == 0:
        print(
            f"Training Alpaca-LoRA model with params:\n"
            f"base_model: {base_model}\n"
            f"data_path: {data_path}\n"
            f"output_dir: {output_dir}\n"
            f"batch_size: {batch_size}\n"
            f"micro_batch_size: {micro_batch_size}\n"
            f"num_epochs: {num_epochs}\n"
            f"learning_rate: {learning_rate}\n"
            f"cutoff_len: {cutoff_len}\n"
            f"val_set_size: {val_set_size}\n"
            f"lora_r: {lora_r}\n"
            f"lora_alpha: {lora_alpha}\n"
            f"lora_dropout: {lora_dropout}\n"
            f"train_on_inputs: {train_on_inputs}\n"
            f"add_eos_token: {add_eos_token}\n"
            f"group_by_length: {group_by_length}\n"
            f"wandb_project: {wandb_project}\n"
            f"wandb_run_name: {wandb_run_name}\n"
            f"wandb_watch: {wandb_watch}\n"
            f"wandb_log_model: {wandb_log_model}\n"
            f"resume_from_checkpoint: {resume_from_checkpoint or False}\n"
            f"prompt template: {prompt_template_name}\n"
        )
    assert (
        base_model
    ), "Please specify a --base_model, e.g. --base_model='huggyllama/llama-7b'"
    gradient_accumulation_steps = batch_size // micro_batch_size

    #prompter = Prompter(prompt_template_name)

    device_map = "auto"
    world_size = int(os.environ.get("WORLD_SIZE", 1))
    ddp = world_size != 1
    if ddp:
        device_map = {"": int(os.environ.get("LOCAL_RANK") or 0)}
        gradient_accumulation_steps = gradient_accumulation_steps // world_size

    # Check if parameter passed or if set within environ
    use_wandb = len(wandb_project) > 0 or (
        "WANDB_PROJECT" in os.environ and len(os.environ["WANDB_PROJECT"]) > 0
    )
    # Only overwrite environ if wandb param passed
    if len(wandb_project) > 0:
        os.environ["WANDB_PROJECT"] = wandb_project
    if len(wandb_watch) > 0:
        os.environ["WANDB_WATCH"] = wandb_watch
    if len(wandb_log_model) > 0:
        os.environ["WANDB_LOG_MODEL"] = wandb_log_model

    tokenizer = AutoTokenizer.from_pretrained(base_model,trust_remote_code=True)
    
    if base_model.find("qwen") != -1 or base_model.find("Qwen") != -1:
        tokenizer.add_special_tokens({"bos_token": "<|im_start|>"})
        tokenizer.add_special_tokens({"eos_token": "<|im_end|>"})
        tokenizer.add_special_tokens({"pad_token": "<|endoftext|>"})
    tokenizer.padding_side = "left"  # Allow batched inference
    def save_model(signal, frame):
        print("\nSaving the model...")
        model.save_pretrained(output_dir)
        sys.exit(0)
    def tokenize(prompt, add_eos_token=True):
        # there's probably a way to do this with the tokenizer settings
        # but again, gotta move fast
        result = tokenizer(
            prompt,
            truncation=True,
            max_length=cutoff_len,
            padding=False,
            return_tensors=None,
        )
        if (
            result["input_ids"][-1] != tokenizer.eos_token_id
            and len(result["input_ids"]) < cutoff_len
            and add_eos_token
        ):
            result["input_ids"].append(tokenizer.eos_token_id)
            result["attention_mask"].append(1)

        result["labels"] = result["input_ids"].copy()

        return result

    def generate_and_tokenize_prompt(data_point):
        full_prompt =data_point["instruction"] + data_point["input"] + data_point["output"]
        tokenized_full_prompt = tokenize(full_prompt)

        return tokenized_full_prompt
    
    print(tokenizer.pad_token_id)
    print(tokenizer.pad_token)
    print(tokenizer.bos_token_id)
    print(tokenizer.bos_token)
    print(tokenizer.eos_token_id)
    print(tokenizer.eos_token)
    if data_path.endswith(".json") or data_path.endswith(".jsonl"):
        data = load_dataset("json", data_files=data_path)
    else:
        data = load_dataset(data_path)
    if val_set_size > 0:
        train_val = data["train"].train_test_split(
            test_size=val_set_size, shuffle=True, seed=42
        )
        train_data = (
            train_val["train"].shuffle().map(generate_and_tokenize_prompt)
        )
        val_data = (
            train_val["test"].shuffle().map(generate_and_tokenize_prompt)
        )
    else:
        train_data = data["train"].shuffle().map(generate_and_tokenize_prompt)
        val_data = None

    model = AutoModelForCausalLM.from_pretrained(base_model,
                                             trust_remote_code=True,
                                             attn_implementation="flash_attention_2",
                                             torch_dtype=torch.bfloat16,
                                             #device_map=device_map,
                                             )


    if resume_from_checkpoint:
        # Check the available weights and load them
        checkpoint_name = os.path.join(
            resume_from_checkpoint, "pytorch_model.bin"
        )  # Full checkpoint
        if not os.path.exists(checkpoint_name):
            checkpoint_name = os.path.join(
                resume_from_checkpoint, "adapter_model.bin"
            )  # only LoRA model - LoRA config above has to fit
            resume_from_checkpoint = (
                False  # So the trainer won't try loading its state
            )
        # The two files above have a different name depending on how they were saved, but are actually the same.
        if os.path.exists(checkpoint_name):
            print(f"Restarting from {checkpoint_name}")
            adapters_weights = torch.load(checkpoint_name)
            set_peft_model_state_dict(model, adapters_weights)
        else:
            print(f"Checkpoint {checkpoint_name} not found")


    if not ddp and torch.cuda.device_count() > 1:
        # keeps Trainer from trying its own DataParallelism when more than 1 gpu is available
        model.is_parallelizable = True
        model.model_parallel = True

    trainer = transformers.Trainer(
        model=model,
        train_dataset=train_data,
        eval_dataset=val_data,
        args=transformers.TrainingArguments(
            per_device_train_batch_size=micro_batch_size,
            gradient_accumulation_steps=gradient_accumulation_steps,
            warmup_steps=0,
            num_train_epochs=num_epochs,
            learning_rate=learning_rate,
            bf16=True,
            logging_steps=50,
            lr_scheduler_type="cosine",
            #optim="adamw_torch",
            optim = "galore_adamw_8bit_layerwise",
            optim_target_modules=[r".*attn.*", r".*mlp.*"],
            
            optim_args="rank=1024, update_proj_gap=500, scale=0.25",
            evaluation_strategy="steps" if val_set_size > 0 else "no",
            save_strategy="steps",
            eval_steps=100 if val_set_size > 0 else None,
            save_steps=200,
            output_dir=output_dir,
            save_total_limit=2,
            load_best_model_at_end=True if val_set_size > 0 else False,
            ddp_find_unused_parameters=False if ddp else None,
            group_by_length=group_by_length,
            report_to="wandb" if use_wandb else None,
            run_name=wandb_run_name if use_wandb else None,
        ),
        data_collator=transformers.DataCollatorForSeq2Seq(
            tokenizer, pad_to_multiple_of=8, return_tensors="pt", padding=True
        ),
    )
    
    
    signal.signal(signal.SIGINT, save_model)
    
    trainer.train()
    model.save_pretrained(output_dir)

    print(
        "\n If there's a warning about missing keys above, please disregard :)"
    )


if __name__ == "__main__":
    fire.Fire(train)

```
result:
```
Map: 100%|█████████████████████████████████████████████████████████████████████████████████████| 2760/2760 [00:20<00:00, 133.85 examples/s]
You are attempting to use Flash Attention 2.0 with a model not initialized on GPU. Make sure to move the model to GPU after initializing it on CPU with `model.to('cuda')`.
Loading checkpoint shards: 100%|█████████████████████████████████████████████████████████████████████████████| 4/4 [00:01<00:00,  3.29it/s]
Using the `WANDB_DISABLED` environment variable is deprecated and will be removed in v5. Use the --report_to flag to control the integrations used for logging result (for instance --report_to none).
/work/jcxy/anaconda3/envs/haolu/lib/python3.10/site-packages/accelerate/accelerator.py:444: FutureWarning: Passing the following arguments to `Accelerator` is deprecated and will be removed in version 1.0 of Accelerate: dict_keys(['dispatch_batches', 'split_batches', 'even_batches', 'use_seedable_sampler']). Please pass an `accelerate.DataLoaderConfiguration` instead: 
dataloader_config = DataLoaderConfiguration(dispatch_batches=None, split_batches=False, even_batches=True, use_seedable_sampler=True)
  warnings.warn(
Activated GaLoRE fine-tuning, depending on your model size and hardware, the training might take a while before starting. Please be patient !
{'loss': 0.8829, 'grad_norm': 0.0, 'learning_rate': 0.001, 'epoch': 0.02}                                                                  
{'loss': 0.7547, 'grad_norm': 0.0, 'learning_rate': 0.001, 'epoch': 0.04}                                                                  
{'loss': 0.7595, 'grad_norm': 0.0, 'learning_rate': 0.001, 'epoch': 0.05}                                                                  
{'loss': 0.7547, 'grad_norm': 0.0, 'learning_rate': 0.001, 'epoch': 0.07}       
```

### PR DESCRIPTION
**Pull request by vasqu**
# What does this PR do?
See #31707 for a detailed rundown. Fixes #31707

Tl;dr: Galore still has issues displaying the correct lr due to the lr scheduler this time. 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@muellerzr @SunMarc @amyeroberts @Minami-su

### CODE REVIEW

**Comment by vasqu** [comment]
Failing tests seem unrelated to me: TF and hub issues.

**Review by muellerzr — COMMENTED** [review]
Thanks, overall this makes sense. Can you add a test in trainer_utils for this by chance? https://github.com/huggingface/transformers/blob/main/tests/trainer/test_trainer_utils.py

**Comment by vasqu** [comment]
@muellerzr Wouldn't it make more sense over here? https://github.com/huggingface/transformers/blob/3345ae733b6f4aeb7204a0f3e646a3cdbaad0023/tests/trainer/test_trainer.py#L1452

I would add two tests:
- A basic learning rate check without a scheduler.
- Cosine with warmup steps and check the lr logs to roughly follow the correct pattern.

Is that reasonable? I'm not sure when I'll have the time tho.

File: tests/trainer/test_trainer.py [hunk 1]
```diff
@@ -1653,6 +1653,84 @@ def test_galore_adafactor_all_linear(self):
...
+                logs[i]["learning_rate"] > logs[i + 1]["learning_rate"]
+                for i in range(len(logs) - 1)
+                if i >= num_warmup_steps - 2
+            ]
+
+            self.assertTrue(all(increasing_lrs))
+            self.assertTrue(all(decreasing_lrs))
+
+            # warm up steps << total steps
+            self.assertTrue(len(decreasing_lrs) > len(increasing_lrs))
```
Comment (vasqu) [inline tests/trainer/test_trainer.py:1732]: Just checking for the general patterns of the cosine scheduler. We could just hardcode the values, but I don't think that's necessary.

Moved the tests in the general trainer tests but could also be moved elsewhere like. Thought it was more appropriate over there.

**Review by muellerzr — APPROVED** [review]
Good job with the test!

cc @amyeroberts for final review

**Review by amyeroberts — COMMENTED** [review]
Thanks for fixing! 

Just a comment on the default LR

File: src/transformers/trainer_pt_utils.py [hunk 2]
```diff
@@ -1379,13 +1380,23 @@ class LayerWiseDummyScheduler(LRScheduler):
...
-        optimizer = LayerWiseDummyOptimizer()
+        optimizer = LayerWiseDummyOptimizer(**kwargs)
         last_epoch = -1
         verbose = False
         super().__init__(optimizer, last_epoch, verbose)
 
     def get_lr(self):
-        return [group["lr"] for group in self.optimizer.param_groups]
+        # default value
+        lrs = [1e-3]
```
[the remark this answers is not in the document]
Comment (vasqu) [reply src/transformers/trainer_pt_utils.py]: The default value is in the dummy optimizer, I'll just save them on the initial creation of the dummy scheduler. This way we won't have the hardcoded value.

File: src/transformers/optimization.py [hunk 3]
```diff
@@ -519,7 +519,7 @@ def scheduler_hook(param):
             if param.requires_grad:
                 param.register_post_accumulate_grad_hook(scheduler_hook)
 
-        return LayerWiseDummyScheduler()
+        return LayerWiseDummyScheduler(optimizer_dict=optimizer_dict, lr=optimizer.defaults.get("lr", 1e-3))
```
[the remark this answers is not in the document]
Comment (vasqu) [reply src/transformers/optimization.py]: It is like a double fallback. Shouldn't be necessary since the dummy optimizer is guaranteed to have a value.

The 1e-3 itself comes from torch galore as their specific defaults.

**Review by amyeroberts — APPROVED** [review]
LGTM - thanks for iterating! 

Rebasing on main should resolve any timeout issues on the CI runs

**Comment by vasqu** [comment]
@amyeroberts One timeout didn't make it through. Is it just my luck? :laughing:

**Comment by amyeroberts** [comment]
@vasqu Just bad luck - although we'll need to look into it on our side why these flaky failures are happening. Thankfully some re-runs worked. Thanks for your patience!
